datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -566,36 +568,57 @@ pub async fn from_substrait_rel(
                 let mut input = LogicalPlanBuilder::from(
                     from_substrait_rel(ctx, input, extensions).await?,
                 );
-                let mut names: HashSet<String> = HashSet::new();
-                let mut exprs: Vec<Expr> = vec![];
-                for e in &p.expressions {
-                    let x =
-                        from_substrait_rex(ctx, e, input.clone().schema(), extensions)
+
+                let mut default_exprs: Vec<Expr> = vec![];
+
+                // By default, a Substrait Project emits all inputs fields
+                let input_schema = input.schema();
+                for index in 0..(input_schema.fields().len()) {
+                    let e =
+                        Expr::Column(Column::from(input_schema.qualified_field(index)));
+                    default_exprs.push(e);
+                }
+                // followed by all expressions
+                for expr in &p.expressions {
+                    let e =
+                        from_substrait_rex(ctx, expr, input.clone().schema(), extensions)
                             .await?;
                     // if the expression is WindowFunction, wrap in a Window relation
-                    if let Expr::WindowFunction(_) = &x {
+                    if let Expr::WindowFunction(_) = &e {
                         // Adding the same expression here and in the project below
                         // works because the project's builder uses columnize_expr(..)
                         // to transform it into a column reference
-                        input = input.window(vec![x.clone()])?
+                        input = input.window(vec![e.clone()])?
                     }
-                    // Ensure the expression has a unique display name, so that project's
-                    // validate_unique_names doesn't fail
-                    let name = x.schema_name().to_string();
-                    let mut new_name = name.clone();
-                    let mut i = 0;
-                    while names.contains(&new_name) {
-                        new_name = format!("{}__temp__{}", name, i);
-                        i += 1;
+                    default_exprs.push(e);
+                }
+
+                // Ensure that all expressions have a unique display name, so that
+                // validate_unique_names does not fail when constructing the project.
+                let mut name_tracker = NameTracker::new();
+
+                let mut final_exprs: Vec<Expr> = vec![];
+                match retrieve_emit_kind(p.common.as_ref()) {
```

Comment:
Isn't this essentially the same thing as `apply_emit_kind`? I get that translating substrait Project into two DF Projects feels odd, but I think that would be more consistent with other Rels w/o treating Project as an exception. imho this is going into optimizer territory... if we do this, why not also try to fold emits into AggregateRel measures as well for example? I think it would be better to simply trust DF optimizer to fold the extra SELECTs afterwards where necessary.

Reply:
oh.. I saw this is addressed in the description already. consider this my 2 cents :laughing:

Reply:
> I get that translating substrait Project into two DF Projects feels odd, but I think that would be more consistent with other Rels w/o treating Project as an exception.

I would actually prefer to output 2 projections, and was actually what I did originally, but that breaks tests like https://github.com/apache/datafusion/blob/1fd6116dd9e1898540b4fbdbba735c4ebacc4227/datafusion/substrait/tests/cases/function_test.rs#L36-L41 because we would need to add the extra projections everywhere. We could do that, but I think that would make writing and reading the tests more burdensome.

> I think it would be better to simply trust DF optimizer to fold the extra SELECTs afterwards where necessary.

I agree as well, but we also don't want to call the optimizer as part of consuming Substrait https://github.com/apache/datafusion/pull/12800

Reply:
Got it, I suppose double projections everywhere can get ugly pretty quickly.

> I agree as well, but we also don't want to call the optimizer as part of consuming Substrait

sure, I meant an optimizer pass after consumer spits out the plan, not as part of a consumer.

Another thing that came to mind just now.. might be a stretch, but bear with me here :laughing: Can't this lead to correctness issues if emit maps the same nondeterministic projection expression to multiple output fields? take random() function for example.. DF would have no way of knowing the original substrait plan asked for a single random value emitted twice rather than two different random values.

Reply:
> Can't this lead to correctness issues if emit maps the same nondeterministic projection expression to multiple output fields?

That is... not really a stretch. This pushes me back towards 2 projects, maybe with an optimization pass to squash duplicate projects as part of the roundtrip tester. I'll give that a shot tomorrow.